src/PostgREST/DbStructure/Relation.hs

```diff
@@ -32,45 +29,36 @@ data Relation = Relation
   , relColumns  :: [Column]
   , relFTable   :: Table
   , relFColumns :: [Column]
-  , relType     :: Cardinality
-  , relLink     :: Link -- ^ Constraint on O2M/M2O, Junction for M2M Cardinality
+  , relCard     :: Cardinality
   }
   deriving (Eq, Generic, JSON.ToJSON)
 
+-- | The relationship cardinality
+-- | https://en.wikipedia.org/wiki/Cardinality_(data_modeling)
+-- TODO: missing one-to-one
+data Cardinality
+  = O2M ConstraintName -- ^ one-to-many cardinality
+  | M2O ConstraintName -- ^ many-to-one cardinality
+  | M2M Junction       -- ^ many-to-many cardinality
+  deriving (Eq, Generic, JSON.ToJSON)
```

Comment:
Ah, you can probably remove one of them. You can see in the html coverage report (when you run `postgrest-coverage` locally), which one it is.

Just had a look - that would be `Eq`. Not need on any of the three types (`Junction`, `Cardinality` or `Relation`).

Reply:
First time using checking the html, that's really cool!

Is there a nix command for that? I just did a `python -mSimpleHTTPServer 8080` on `/coverage`.

Reply:
Fair question! We have nix tools for everything. Except that.

I just open the index file directly in the browser via file manager. Works just as fine.

Reply:
Actually. Exactly for that case I have added the `file://` URI to the CLI output. My terminal has clickable-URI support... so I should not even use a file manager :)

Reply:
Cool. I'll have to level up my terminal game then :)

>  that would be Eq. Not need on any of the three types (Junction, Cardinality or Relation).

Can't seem to do that:

```haskell
src/PostgREST/Request/DbRequestBuilder.hs:306:61: error:
    • No instance for (Eq Relation) arising from a use of ‘delete’
    • In the second argument of ‘(:)’, namely ‘delete tn forest’
      In the second argument of ‘Node’, namely
        ‘(addProperty f (remainingPath, a) tn : delete tn forest)’
      In the expression:
        Node rn (addProperty f (remainingPath, a) tn : delete tn forest)
    |
306 |     Just tn -> Node rn (addProperty f (remainingPath, a) tn:delete tn forest)
```

Reply:
Weird because I also see the same "never executed" notice on the Eqs, on the coverage html.

Reply:
Well, if the only place where `Eq` is used is uncovered, too.... then we need it of course :)

I should have been more precise. It's possible that one of the 3 `Eq`s is unused - you can find out by removing. If no error, it can be safely removed.

Reply:
> Well, if the only place where `Eq` is used is uncovered, too.... then we need it of course :)

Which does not seem to be the case here, I think. That line you referenced, does seem to be covered.

I think the only thing that can happen here is, that `delete` is only ever run with an empty list here. That's why it never calls `Eq` internally. But it still depends on it.

Reply:
I don't pretend to understand what the `addProperty` function does, so I don't know whether that makes any sense, whether we're missing a test-case here or whether this is kind of an impossible branch.

Reply:
> Well, if the only place where Eq is used is uncovered, too...

Ah, but the line where it fails it doesn't show as uncovered:

https://github.com/PostgREST/postgrest/blob/1f206a560bedf0d7971e9db5e475c37ec311acda/src/PostgREST/Request/DbRequestBuilder.hs#L307

> I should have been more precise. It's possible that one of the 3 Eqs is unused - you can find out by removing. If no error, it can be safely removed.

No dice for Eq. I remove it on Cardinality or Junction and Relation fails compiling, I remove the Eq from Relation and it gives the [above error](https://github.com/PostgREST/postgrest/pull/1818#discussion_r615012352). Maybe the coverage "never executed" is related to the record fields?

Anyway, I guess I'll just leave it as it is.

Reply:
> I don't pretend to understand what the addProperty function does, so I don't know whether that makes any sense, whether we're missing a test-case here or whether this is kind of an impossible branch.

No, that part is related to [embedded filters](https://postgrest.org/en/stable/api.html#embedded-filters). When having `&roles.character=in.(Chico,Harpo,Groucho)`, addProperty adds the `.character` part if it's a valid embed or drops it silently(ref: https://github.com/PostgREST/postgrest/issues/1771). Code definitely needs cleaning.

Reply:
> No dice for Eq. I remove it on Cardinality or Junction and Relation fails compiling, I remove the Eq from Relation and it gives the [above error](https://github.com/PostgREST/postgrest/pull/1818#discussion_r615012352). Maybe the coverage "never executed" is related to the record fields?

Ah, of course that makes sense. If we want `Eq` on `Relation` we need it on both the others, too - otherwise a comparison on `Relation` could never be done.

The coverage line is for sure not related to the record fields, I had seen many of those already.

I just noted the `delete` is actually part of a recursive call. Maybe the problem is that, we don't have a test for an embedded filter on a nested embedding? Something like `films.roles.character=in...`. Eh, I keep questioning myself whether that's actually possible to do, because I never find any docs on it. But we **can** do nested embeddings, right?

Reply:
> Maybe the problem is that, we don't have a test for an embedded filter on a nested embedding? Something like films.roles.character=in...

True, I think we don't have such test.

> Eh, I keep questioning myself whether that's actually possible to do, because I never find any docs on it. But we can do nested embeddings, right?

Yes, correct. To an N depth, which always reminds me of https://github.com/PostgREST/postgrest/issues/249#issuecomment-334630205. I've put an idea on https://github.com/PostgREST/postgrest/pull/1710#issuecomment-808788905 (at the bottom) to solve that one.